src/radical/pilot/agent/executing/popen.py

```diff
@@ -122,222 +104,312 @@ def command_cb(self, topic, msg):
+        # for all ranks, then the individual `pre_rank` directives are executed,
+        # and then, after all ranks are synchronized, finally the task ranks
+        # begin to run.
+        #
+        # The scripts thus show the following structure:
+        #
+        # Launcher Script (`task.000000.launch.sh`):
+        # ----------------------------------------------------------------------
+        # #!/bin/sh
+        #
+        # # `pre_launch` commands
+        # date > data.input
+        #
+        # # launcher specific environment setup
+        # module load openmpi
+        #
+        # # launch the task script
+        # mpirun -n 4 ./task.000000.exec.sh
+        # ----------------------------------------------------------------------
+        #
+        # Task Script (`task.000000.exec.sh`)
+        # ----------------------------------------------------------------------
+        # #!/bin/sh
+        #
+        # # clean launch environment
+        # module unload mpi
+        #
+        # # task environment setup (`pre_exec`)
+        # module load gromacs
+        #
+        # # rank specific setup (`pre_rank`)
+        # touch task.000000.ranks
+        # if test "$MPI_RANK" = 0; then
+        #   export CUDA_VISIBLE_DEVICES=0
+        #   export OENMP_NUM_THREADS=2
+        #   export RANK_0_VAR=foo
+        #   echo 0 >> task.000000.ranks
+        # fi
+        #
+        # if test "$MPI_RANK" = 1; then
+        #   export CUDA_VISIBLE_DEVICES=1
+        #   export OENMP_NUM_THREADS=4
+        #   export RANK_1_VAR=bar
+        #   echo 1 >> task.000000.ranks
+        # fi
+        #
+        # # synchronize ranks
+        # while $(cat task.000000.ranks | wc -l) != $MPI_RANKS; do
+        #   sleep 1
+        # done
+        #
+        # # run the task
+        # mdrun -i ... -o ... -foo ... 1> task.000000.$MPI_RANK.out \
+        #                              2> task.000000.$MPI_RANK.err
+        #
+        # # now do the very same stuff for the `post_rank` and `post_exec`
+        # # directives
+        # ...
+        #
+        # ----------------------------------------------------------------------
+        #
+        # NOTE: MongoDB only accepts string keys, and thus the rank IDs in
+        #       pre_rank and post_rank dictionaries are rendered as strings.
+        #       This should be changed to more intuitive integers once MongoDB
+        #       is phased out.
+
+        tid  = task['uid']
+        td   = task['description']
+        sbox = task['task_sandbox_path']
+
+        launcher = self.find_launcher(task)
+
+        if not launcher:
+            raise RuntimeError('no launcher foud for task %s' % tid)
+
+        self._log.debug('Launching task with %s', launcher.name)
+
+        launch_script = '%s.launch.sh' % tid
+        exec_script   = '%s.exec.sh'   % tid
+
+        ru.rec_makedir(sbox)
+
+        with open('%s/%s' % (sbox, launch_script), 'w') as fout:
```

Comment:
flag this part to have one `write` call, same as above

Reply:
Same, done.